### DISCUSSION TITLE
Hints about tools for generating typed Python clients

### DISCUSSION THREAD

**Discussion by fgebhart** [actix/actix-web#3487] [Q&A]
Hi there 👋 

we built an API application using actix and have a bunch of python applications that interact with the API. As the interface is subject to frequent changes, we are looking for some tooling that allows us to generate some Python code that implements the interface and types of the endpoints of our API application, so we can actually validate/enforce the python applications to be in sync with the actix API app.

One potential solution to this I could imaging is generating OpenAPI specs using e.g. [apistos](https://github.com/netwo-io/apistos) or [paperclip](https://github.com/paperclip-rs/paperclip) and subsequently generate pydantic classes based on the OpenAPI specs.

Haven't really tried that approach yet and hoped to maybe learn about a few other options before implementing the one option I'm aware of.

Thanks in advance!

**Comment by Samielakkad** [comment]
The OpenAPI route is the one I would reach for first. Actix itself does not have a single built-in source of truth for "all handlers + request/response types" that can generate Python clients directly, so the usual stable pipeline is:

```text
Actix handlers/types -> OpenAPI document -> generated Python client/models
```

A practical setup looks like this:

1. Generate OpenAPI from the Rust side

Common choices are:

- `utoipa`: probably the most actively used option at the moment; you annotate request/response types and paths.
- `apistos`: nice if you want a more integrated typed route style.
- `paperclip`: historically popular, but I would check its current maintenance/status before starting a new project with it.

For new code I would usually start with `utoipa`, especially if you are fine adding explicit annotations to make the contract clear.

2. Generate the Python client

Good Python-side options:

- `openapi-python-client`: generates a typed client package, with models and endpoint functions.
- `datamodel-code-generator`: great for generating Pydantic models from OpenAPI/JSON Schema, especially if you want to write the HTTP client layer yourself.
- OpenAPI Generator: broader ecosystem support, but the generated Python can be heavier depending on your preferences.

If your Python apps mostly need a ready-to-use client, try `openapi-python-client` first. If you mostly want shared Pydantic models and custom request code, try `datamodel-code-generator`.

3. Add a contract check in CI

This is the part that prevents drift. For example:

- generate/export `openapi.json` from the Actix app;
- compare it against the committed spec, or regenerate the Python client in CI;
- fail CI when the generated diff is not committed.

That gives you a clear workflow: when the Actix API changes, the OpenAPI spec changes, then the Python client changes in the same PR.

One recommendation: do not rely only on inferred schemas if the API is changing often. Put request/response DTOs in dedicated structs, derive `Serialize`/`Deserialize`/schema traits on those, and use those same DTOs in handlers. That keeps your public API contract separate from internal database/domain structs and makes the generated Python types much cleaner.
